Optimize Map/Set

sctice opened the pull request:
`Map` as previously implemented used a linear search to find keys, resulting in O(n) put/delete performance, which isn't what one expects from a map (or a `Set`, which uses `Map` under the hood). It's slower than we find acceptable for general use. This appears to have been intentional, to keep the implementation simple, but that makes the class more of a reference implementation than a polyfill, and we actually want a polyfill, to allow us to use the `Ds` interface while we figure out the source of a segfault in the extension.

To improve the situation, we implement `Map` using two parallel arrays: one to maintain order (required by the interface) and another to provide a lookup table by key. We store the same `PairRef` (a container for key and value) in both tables so that finding a pair in either array is sufficient to find it in the other one, even as the order array changes.

This approach leads to a Map/Set implementation that's _roughly_ 8 times slower than using `SplObjectStorage` for `new stdClass` instances and 5 times slower than using native arrays directly for integers, and uses roughly 3 times more memory across the board. For example, storing around 32K integers eats up ~2M for a native array and ~6M in our Map, and takes 5–10ms for the native array versus ~25ms for our Map. It's not surprising that the PHP implementation is slower and more memory hungry than the C implementations, but the new Map/Set polyfill is _significantly_ faster than the previous one (which had made it less than 1% through the benchmark after 10 minutes where the new implementation finishes the benchmark in about 2 minutes) for a modest increase in memory usage (because we add a lookup table to the ordered list of pairs).

In order to test on PHP 8 (and on a machine with the extension enabled), I had to make some minor changes to the test config and the test suite. Note that we use a fork of the test repo, and we use the `php8` branch on that fork.

---

After these changes, we still need to implement custom serialization logic on the polyfill so that its serialized data is compatible with serialized data created by the extension.

qa_req 0

Connects iFixit/ifixit#38493

sctice-ifixit committed:
Improve Map get/remove performance

Rather than keeping a list of pairs and doing a linear search any time we need
to find a key, use a hashing strategy similar to what the extension uses in
order to maintain a lookup table by key hash. Maintain a parallel ordered list
for operations that need to respect order (e.g., iteration, reverse, sort, and
so on). In order to keep the two parallel structure in sync, wrap each pair in
a "PairRef" that keeps track of the pair's position in the internal ordered
list, and update the PairRefs when changing the order (e.g., after a sort).

sctice-ifixit committed:
Increase memory limit on test runs

Previously, I believe we were relying on the default (or whatever the php.ini
specifies), which was too little in my test environment.

sctice-ifixit committed:
Use forked tests for PHP 8

sctice-ifixit committed:
Ignore system php.ini

This prevents the Ds extension from being enabled, so the tests will use the
polyfill.

sctice-ifixit committed:
Map: Only use array for collision list of size 2+

Arrays eat up a lot of memory, and there's also overhead involved in
initializing, growing, and iterating over them. Since we hope our hashing
strategy will result in relatively few collisions, we can avoid a lot of array
allocation and iteration by first associating a PairRef with a hash key, only
"upgrading" to an array once we have an actual collision. We pay in terms of
complexity because insertion, lookup, and deletion have to handle the two cases
for a value associated with a hash key: a PairRef or an array of PairRefs.

After making this change and running the benchmark, it appears that this more
complex approach is somewhat faster and about half as memory-intensive as the
previous (always use an array) approach. The main goal here is to reduce memory
usage for relatively small data sets (up to a few thousand values), so this
seems like a definite win.

sctice-ifixit committed:
Further reduce Map memory usage

Rather than wrapping a Pair in a PairRef, capture the relevant Pair
functionality in PairRef and add the ability to export to a Pair. This saves
a bit on Map storage because we now create one PairRef per value instead of
both a Pair and a PairRef, and it's a bit faster, presumably because we're
allocating less and doing less dereferencing to get at pair keys/values.

It's relatively rare that we actually want to get Pair instances out of the
collection, and in the cases where we do, we always want a copy. Previously,
some methods (e.g., first()) were returning a Pair directly instead of a copy,
but I believe that was a mistake. I think the extension makes a new Pair with
the internal key and value in these cases, and that strikes me as more
reasonable.

sctice-ifixit committed:
Map: Implement filter as loop over pairs

This is the same structure we use for other iterations over all pairs.

danielcliu commented on src/Map.php:
```diff
@@ -399,14 +487,22 @@ public function pairs(): Sequence
+            if (!array_key_exists($hash, $this->table)) {
+                $this->table[$hash] = [$pairRef];
+            } else {
+                $this->table[$hash][] = $pairRef;
```
I don't really understand why this if statement is needed, shouldn't we iterate the $hash or something instead

sctice replied:
Yep! We have to handle collisions. We're only switching to an array when there's an actual collision in order to save memory (since arrays allocate a lot of memory and collisions should be rare).

danielcliu commented on src/Map.php:
```diff
@@ -271,7 +362,7 @@ public function hasValue($value): bool
     public function count(): int
     {
-        return count($this->pairs);
+        return count($this->pairRefs);
```
count() doesn't count recursively i think so because we can have multiple values in a pairRef index couldn't this be inaccurate?

danielcliu commented on src/Map.php:
```diff
@@ -291,7 +382,9 @@ public function filter(callable $callback = null): Map
         $filtered = new self();
 
-        foreach ($this as $key => $value) {
+        foreach ($this->pairRefs as $pairRef) {
```
isn't pairRef an array? It wont have a key and value right?

danielcliu commented on src/Map.php:
```diff
@@ -210,34 +283,52 @@ private function keysAreEqual($a, $b): bool
+        }
+
+        $hashMatch = $this->table[$hash];
+        if ($hashMatch instanceof PairRef) {
```
is there any instance where it wont be an array?

sctice replied:
In order to save on memory, we start out by associating each hash key with just a `PairRef`. We only convert the value associated with a hash key to `list<PairRef>` once we encounter a collision (two different values that hash to the same thing); at that point we'll put the first value and the new value into a new array, which will be size 2. This saves on memory because, in PHP, initializing an array with one thing in it takes up a lot of extra memory compared to just the one thing on its own, and since we expect collisions to be rare, we don't often have to change up from the single `PairRef` to the `list<PairRef>`. This also ends up being faster, I believe because we're avoiding a bunch of allocations for the arrays.

danielcliu replied:
ok i see i was looking at an old commit and you change it

jarstelfox commented on src/Map.php:
```diff
@@ -188,6 +205,62 @@ public function diff(Map $map): Map
+     *
+     * @return mixed The string or integer hash value.
+     *
+     * @psalm-return string|int
```
Minor but you might be able to teach psalm about when int vs string is returned

sctice replied:
I don't think it matters, since we're just trying to get an `array-key`.

jarstelfox commented:
CR 🌵 Most of the changes are easy to CR, however, I do want to point out the get_hash function is a bit wild.

danielcliu commented on src/Map.php:
```diff
@@ -235,7 +321,8 @@ private function lookupKey($key)
      */
     private function lookupValue($value)
```
shouldn't we be able to use the hash here for a quicker lookup?

danielcliu commented on src/Map.php:
```diff
@@ -25,11 +25,25 @@ final class Map implements Collection, \ArrayAccess
+     *
+     * @psalm-var array<array-key, list<PairRef>>
+     */
+    private $table = [];
```
i feel like this coudl potentiallly be reduced to PairRef->key and then use that to lookup in the pairRefs table

sctice replied:
I'm not quite following. The `pairRefs` array doesn't have (stable, meaningful) keys. It maintains the order of elements in the map.

danielcliu replied:
ah i see i thought the keys were the hashes for some reason

danielcliu commented on src/Map.php:
```diff
@@ -399,14 +487,22 @@ public function pairs(): Sequence
             $this->checkCapacity();
-            $this->pairs[] = new Pair($key, $value);
+            $pair = new Pair($key, $value);
+            $pairIndex = $this->nextPairIndex++;
```
is there a reason we cant just `$this->pairRefs[] = $pairRef?`

sctice replied:
We need the index to go into the `PairRef` constructor so that we can keep it up to date when the map order changes, e.g., after a reverse or sort. We can't get the index from `count($this->pairs)` because the array can have holes in it from deletes.

danielcliu commented:
if memory is an issue we could not use arrays at all to hold the pairRefs in table and do the traditional iteration (like if there is a collision you have a backoff of index to jump after and look there ithink). You pay for this with deletes i think cause you have to reorganize the table every time, but its an option i think (although with phps not quite dictionary lists maybe it isn't).

Anyway some small comments but CR ✌️

sctice commented:
> if memory is an issue we could not use arrays at all to hold the pairRefs in table and do the traditional iteration (like if there is a collision you have a backoff of index to jump after and look there ithink). You pay for this with deletes i think cause you have to reorganize the table every time, but its an option i think (although with phps not quite dictionary lists maybe it isn't).

If we hashed everything to an integer, then I believe we could use an `SplFixedArray` as you suggest, but that would require figuring out our own balanced hashing algorithm instead of relying on PHP to do that for us, appropriately growing the array as we add new elements, and it'd probably be quite a bit slower because we'd be doing it in interpreted PHP code instead of having PHP do it in compiled C. Although there's some complexity in dealing with a union type (`PairRef|list<PairRef>`), I think it strikes a reasonable balance between memory and time.

k0rvusk0r4x commented:
Ideas on qaing this?

sctice-ifixit committed:
Add decription to composer.json

sctice-ifixit committed:
Update test scripts to work with GH actions

For `test`, just run PHPUnit, which will use the extension if it's present (so
don't use it when you expect the extension to be present).

For `test-no-ini`, specify the `dom` extension, which PHPUnit requires. You can
use this script even when the extension is installed because we don't tell PHP
to load it.

sctice commented:
> Ideas on qaing this?

I think passing unit tests should be sufficient, and I've set up an action to run the tests.

jarstelfox commented:
CR 🌵

danielcliu commented:
CR ✌️

sctice commented:
> however, I do want to point out the get_hash function is a bit wild.

Haha, yeah. Anything in particular you'd like to improve about it? My assumption in writing it is that our main objectives, in order are:

- Convert _any_ PHP value to either a string or an integer so that we can use it as a key in a PHP array.
- Try to avoid very long string keys, because I believe those will require an equality check on each internal lookup.
- Let PHP take hashing from there (i.e., don't worry about generating uniformly-distributed hash keys, since PHP is going to hash our "hash keys" again when we index into `$table`).
